### PR TITLE
refactor: fix(secgroup): The security group and rules support eps authorization

### LINKED ISSUE #1912: [EPS] The security group does not support eps authorization
Hi team,

The resources of the security group have some problems:

- The description of security group cannot be updated correctly.
- The security group name cannot be updated correctly.
- The security group rules cannot be created correctly.
- The list of security groups cannot be obtained correctly.

Best regards!

### PR DESCRIPTION
**Pull request by Lance52259**
**What this PR does / why we need it**:
Currently, the openstack API does not support enterprise project authorization. Need to switch to the HuaweiCloud self-developed API and update the logic of the query URL creation about `enterprise_project_id`.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #1912

**Special notes for your reviewer**:

**Release note**:

```release-note
1. replace the OpenStack API with the HuaweiCloud v3 API. 
2. update the splice logic of query URL about List method.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

data.huaweicloud_networking_secgroup
```
make testacc TEST='./huaweicloud' TESTARGS='-run=TestAccNetworkingSecGroupV3DataSource'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccNetworkingSecGroupV3DataSource -timeout 360m -parallel 4
=== RUN   TestAccNetworkingSecGroupV3DataSource_basic
=== PAUSE TestAccNetworkingSecGroupV3DataSource_basic
=== RUN   TestAccNetworkingSecGroupV3DataSource_byID
=== PAUSE TestAccNetworkingSecGroupV3DataSource_byID
=== CONT  TestAccNetworkingSecGroupV3DataSource_basic
=== CONT  TestAccNetworkingSecGroupV3DataSource_byID
--- PASS: TestAccNetworkingSecGroupV3DataSource_basic (113.69s)
--- PASS: TestAccNetworkingSecGroupV3DataSource_byID (118.30s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       119.336s
```
huaweicloud_networking_secgroup
```
make testacc TEST='./huaweicloud' TESTARGS='-run=TestAccNetworkingV3SecGroup'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccNetworkingV3SecGroup -timeout 360m -parallel 4
=== RUN   TestAccNetworkingV3SecGroup_basic
=== PAUSE TestAccNetworkingV3SecGroup_basic
=== RUN   TestAccNetworkingV3SecGroup_withEpsId
=== PAUSE TestAccNetworkingV3SecGroup_withEpsId
=== RUN   TestAccNetworkingV3SecGroup_noDefaultRules
=== PAUSE TestAccNetworkingV3SecGroup_noDefaultRules
=== CONT  TestAccNetworkingV3SecGroup_basic
=== CONT  TestAccNetworkingV3SecGroup_noDefaultRules
=== CONT  TestAccNetworkingV3SecGroup_withEpsId
--- PASS: TestAccNetworkingV3SecGroup_withEpsId (91.83s)
--- PASS: TestAccNetworkingV3SecGroup_noDefaultRules (94.94s)
--- PASS: TestAccNetworkingV3SecGroup_basic (137.39s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       137.557s
```
huaweicloud_networking_secgroup_rule
```
make testacc TEST='./huaweicloud' TESTARGS='-run=TestAccNetworkingV2SecGroupRule'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccNetworkingV2SecGroupRule -timeout 360m -parallel 4
=== RUN   TestAccNetworkingV2SecGroupRule_basic
=== PAUSE TestAccNetworkingV2SecGroupRule_basic
=== RUN   TestAccNetworkingV2SecGroupRule_remoteGroup
=== PAUSE TestAccNetworkingV2SecGroupRule_remoteGroup
=== RUN   TestAccNetworkingV2SecGroupRule_lowerCaseCIDR
=== PAUSE TestAccNetworkingV2SecGroupRule_lowerCaseCIDR
=== CONT  TestAccNetworkingV2SecGroupRule_basic
=== CONT  TestAccNetworkingV2SecGroupRule_lowerCaseCIDR
=== CONT  TestAccNetworkingV2SecGroupRule_remoteGroup
--- PASS: TestAccNetworkingV2SecGroupRule_lowerCaseCIDR (123.40s)
--- PASS: TestAccNetworkingV2SecGroupRule_basic (124.13s)
--- PASS: TestAccNetworkingV2SecGroupRule_remoteGroup (125.57s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       126.848s
```
loadbalancer with security group
```
make testacc TEST='./huaweicloud' TESTARGS='-run=TestAccLBV2LoadBalancer_secGroup'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccLBV2LoadBalancer_secGroup -timeout 360m -parallel 4
=== RUN   TestAccLBV2LoadBalancer_secGroup
=== PAUSE TestAccLBV2LoadBalancer_secGroup
=== CONT  TestAccLBV2LoadBalancer_secGroup
--- PASS: TestAccLBV2LoadBalancer_secGroup (156.38s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       156.543s
```

### CODE REVIEW

File: docs/resources/networking_secgroup_rule.md [hunk 1]
```diff
@@ -50,11 +49,9 @@ The following arguments are supported:
   __icmp__ and __icmpv6__. If omitted, the protocol means that all protocols are supported.
   This is required if you want to specify a port range. Changing this creates a new security group rule.
 
-* `port_range_min` - (Optional, Int, ForceNew) Specifies the lower part of the allowed port range, valid integer value
-  needs to be between 1 and 65535. Changing this creates a new security group rule.
-
-* `port_range_max` - (Optional, Int, ForceNew) Specifies the higher part of the allowed port range, valid integer value
-  needs to be between 1 and 65535. Changing this creates a new security group rule.
+* `ports` - (Optional, String, ForceNew) Specifies the allowed port value range, which supports single port (80),
```
Comment (niuzhenguo) [inline docs/resources/networking_secgroup_rule.md:52]: I'd like to make it compatible with `port_range_min` and `port_range_max`, it's OK to just directly update here but need to be compatible in codes. `port_range_min` and `port_range_max` can be converted to singile port (80) or  continuous port (1-30).

**Comment by niuzhenguo** [comment]
be careful about the `breaking change`

File: docs/resources/networking_secgroup_rule.md (hunk 1 above)
Comment (Lance52259) [reply docs/resources/networking_secgroup_rule.md:52]: OFC

File: huaweicloud/config/endpoints.go [hunk 2]
```diff
@@ -176,6 +176,10 @@ var allServiceCatalog = map[string]ServiceCatalog{
 		Version:          "v2.0",
 		WithOutProjectID: true,
 	},
+	"networkv3": {
```
Comment (ShiChangkuo) [inline huaweicloud/config/endpoints.go]: s/networkv3/vpcv3

File: huaweicloud/data_source_huaweicloud_networking_secgroup.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/chnsz/golangsdk/openstack/networking/v3/security/groups"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/common"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/config"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils/fmtp"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils/logp"
+)
+
+func DataSourceNetworkingSecGroupV3() *schema.Resource {
```
Comment (ShiChangkuo) [inline huaweicloud/data_source_huaweicloud_networking_secgroup.go]: the name should be without version

File: huaweicloud/data_source_huaweicloud_networking_secgroup.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+	networkingClient, err := config.NetworkingV3Client(GetRegion(d, config))
+	if err != nil {
+		return fmtp.DiagErrorf("Error creating HuaweiCloud networking client: %s", err)
+	}
+
+	listOpts := groups.ListOpts{
+		ID:   d.Get("secgroup_id").(string),
+		Name: d.Get("name").(string),
+	}
+	if epsId := common.GetEnterpriseProjectID(d, config); epsId != "" {
```
Comment (ShiChangkuo) [inline huaweicloud/data_source_huaweicloud_networking_secgroup.go:67]: `enterprise_project_id` is not defined in schema

File: huaweicloud/resource_huaweicloud_networking_secgroup.go [hunk 5]
```diff
@@ -102,108 +100,107 @@ func ResourceNetworkingSecGroupV2() *schema.Resource {
...
 	if mErr.ErrorOrNil() != nil {
-		return mErr
+		return diag.FromErr(mErr)
 	}
 
 	return nil
 }
 
-func flattenSecurityGroupRules(secGroup *securitygroups.SecurityGroup) []map[string]interface{} {
+func flattenSecurityGroupRules(secGroup *groups.SecurityGroup) []map[string]interface{} {
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_networking_secgroup.go]: please merge it with `flattenSecurityGroupDataRules` function

File: huaweicloud/resource_huaweicloud_networking_secgroup_rule.go [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+	meta interface{}) diag.Diagnostics {
+	config := meta.(*config.Config)
+	client, err := config.NetworkingV3Client(GetRegion(d, config))
+	if err != nil {
+		return fmtp.DiagErrorf("Error creating HuaweiCloud networking v3 client: %s", err)
+	}
+
+	protocol := d.Get("protocol").(string)
+	if _, ok := d.GetOk("ports"); ok && protocol == "" {
+		return fmtp.DiagErrorf("A protocol must be specified when using port_range_min and port_range_max")
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_networking_secgroup_rule.go]: the message is misleading.
also you can use `Required` in schema to check it.

File: huaweicloud/resource_huaweicloud_networking_secgroup_rule.go [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+		RemoteIpPrefix:  d.Get("remote_ip_prefix").(string),
+		Protocol:        protocol,
+	}
+
+	if v, ok := d.GetOk("direction"); ok {
+		opt.Direction = v.(string)
+	}
+	if v, ok := d.GetOk("ethertype"); ok {
+		opt.Ethertype = v.(string)
+	}
```
Comment (ShiChangkuo) [inline huaweicloud/resource_huaweicloud_networking_secgroup_rule.go]: GetOk will always be **true**, as they are required parameters

Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_networking_secgroup_rule.go]: yeah, both **direction** and **ethertype** behavior are required.

File: huaweicloud/resource_huaweicloud_networking_secgroup_rule.go (hunk 6 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_networking_secgroup_rule.go]: done.
1. The **port_range_min** and **port_range_max** needs to be set.
2. The **ports** is conflict with **port_range_min** and **port_range_max**.
3. The **protocol** is required if **ports** or **port_range_min** or **port_range_max** is set.

File: huaweicloud/resource_huaweicloud_networking_secgroup.go (hunk 5 above)
Comment (Lance52259) [reply huaweicloud/resource_huaweicloud_networking_secgroup.go]: merged

File: huaweicloud/data_source_huaweicloud_networking_secgroup.go (hunk 3 above)
Comment (Lance52259) [reply huaweicloud/data_source_huaweicloud_networking_secgroup.go]: All version labels have been removed from methods.

File: huaweicloud/config/endpoints.go (hunk 2 above)
Comment (Lance52259) [reply huaweicloud/config/endpoints.go]: done

**Comment by Lance52259** [comment]
Now, only the update of the parameter **tenant_id** is a breaking change.

File: huaweicloud/data_source_huaweicloud_networking_secgroup.go (hunk 4 above)
Comment (Lance52259) [reply huaweicloud/data_source_huaweicloud_networking_secgroup.go:67]: Missing parameter has been added.

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/lb' TESTARGS='-run=TestAccLBV2LoadBalancer_secGroup'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/lb -v -run=TestAccLBV2LoadBalancer_secGroup -timeout 360m -parallel 4
=== RUN   TestAccLBV2LoadBalancer_secGroup
=== PAUSE TestAccLBV2LoadBalancer_secGroup
=== CONT  TestAccLBV2LoadBalancer_secGroup
--- PASS: TestAccLBV2LoadBalancer_secGroup (176.79s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/lb        177.776s
```

**Pull request merged**